File: src/test/java/com/seikim/passort/member/application/MemberServiceTest.java
package com.seikim.passort.member.application;

import com.seikim.passort.exception.PassportErrorCode;
import com.seikim.passort.exception.PassportException;
import com.seikim.passort.passport.Passport;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.transaction.annotation.Transactional;

import java.time.LocalDateTime;

import static org.assertj.core.api.Assertions.assertThatCode;
import static org.assertj.core.api.Assertions.assertThatThrownBy;

@Transactional
@SpringBootTest
class MemberServiceTest {

    @Autowired
    private MemberService memberService;

    private Passport passport;

    @BeforeEach
    void setup() {
        passport = Passport.builder()
                .memberId(1L)
                .userAgent("USER_AGENT")
                .path("PATH")
                .expirationTime(LocalDateTime.now().plusSeconds(5))
                .build();
    }

    @DisplayName("Passport 검증에 성공한다.")
    @Test
    void validatePassportSuccessTest() {
        assertThatCode(() -> memberService.doing(passport))
                .doesNotThrowAnyException();
    }

    @DisplayName("이미 사용된 Passport 검증에 성공한다.")
    @Test
    void validatePassportAlreadyUsedFailed() {
        memberService.doing(passport);
        assertThatThrownBy(() -> memberService.doing(passport))
                .isInstanceOf(PassportException.class)
                .hasMessage(PassportErrorCode.ALREADY_USED_PASSPORT.getMessage());
    }

    @DisplayName("만료된 Passport 검증에 성공한다.")
    @Test
    void validatePassportExpireTimeFailed() {
        Passport expiredPassport = Passport.builder()
                .memberId(1L)
                .userAgent("USER_AGENT")
                .path("PATH")
                .expirationTime(LocalDateTime.now().minusSeconds(10L))
                .build();

        assertThatThrownBy(() -> memberService.doing(expiredPassport))
                .isInstanceOf(PassportException.class)
                .hasMessage(PassportErrorCode.EXPIRED_PASSPORT.getMessage());
    }
}

File: kotlin-basic/src/test/kotlin/com/seikim/kotlinbasic/member/service/MemberServiceTest.kt
package com.seikim.kotlinbasic.member.service

import com.seikim.kotlinbasic.member.domain.MemberDataBase
import com.seikim.kotlinbasic.member.fixture.MemberFixture
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.assertAll
import org.junit.jupiter.params.ParameterizedTest
import org.junit.jupiter.params.provider.EnumSource
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.test.context.SpringBootTest

@SpringBootTest
class MemberServiceTest(
    @Autowired val memberService: MemberService,
    @Autowired val memberDataBase: MemberDataBase
) {
    @ParameterizedTest
    @EnumSource(MemberFixture::class)
    fun `회원 저장에 성공한다`(fixture: MemberFixture) {
        // given
        // when
        val savedMember = memberService.save(fixture.memberName, fixture.email)

        // then
        assertAll(
            { assertThat(savedMember.id).isGreaterThan(0) },
            { assertThat(savedMember.name).isEqualTo(fixture.memberName) },
            { assertThat(savedMember.email).isEqualTo(fixture.email) }
        )
    }

    @ParameterizedTest
    @EnumSource(MemberFixture::class)
    fun `사용자 조회에 성공한다`(fixture: MemberFixture) {
        // given
        val savedMember = memberDataBase.save(fixture.생성한다())

        // when
        val findMember = memberService.findById(savedMember.id)

        // then
        assertThat(findMember).isEqualTo(savedMember)
    }
}